KAFKA-7800: Dynamic log levels admin API (KIP-412)

stanislavkozlovski opened the pull request:
<!--
Is there any breaking changes?  If so this is a major release, make sure '#major' is in at least one
commit message to get CI to bump the major.  This will prevent automatic down stream dependency
bumping / consuming.  For more information about semantic versioning see: https://semver.org/


Suggested PR template: Fill/delete/add sections as needed. Optionally delete any commented block.
-->
What
----
<!--
Briefly describe **what** you have changed and **why**.
Optionally include implementation strategy.
-->

References
----------
[**KIP-412**](https://cwiki.apache.org/confluence/display/KAFKA/KIP-412%3A+Extend+Admin+API+to+support+dynamic+application+log+levels)
[**KAFKA-7800**](https://issues.apache.org/jira/browse/KAFKA-7800)
[**Discussion Thread**](http://mail-archives.apache.org/mod_mbox/kafka-dev/201901.mbox/%3CCANZZNGyeVw8q%3Dx9uOQS-18wL3FEmnOwpBnpJ9x3iMLdXY3gEug%40mail.gmail.com%3E)
[**Vote Thread**](http://mail-archives.apache.org/mod_mbox/kafka-dev/201902.mbox/%3CCANZZNGzpTJg5YX1Gpe5S%3DHSr%3DXGvmxvYLTdA3jWq_qwH-UvorQ%40mail.gmail.com%3E)

<!--
Copy&paste links: to Jira ticket, other PRs, issues, Slack conversations...
For code bumps: link to PR, tag or GitHub `/compare/master...master`
-->

Test&Review
------------
Test cases covered:
* DescribeConfigs
* Alter the log level with and without validateOnly, validate the results with DescribeConfigs

Open questions / Follow ups
--------------------------
If you're a reviewer, I'd appreciate your thoughts on these questions I have open:
1. Should we add synchronization to the Log4jController methods? - Seems like we don't get much value from it
2. Should we instantiate a new Log4jController instead of it having static methods? - All operations are stateless, so I thought static methods would do well
3. A logger which does not have a set value returns "null" (as seen in the unit tests). Should we just return the Root logger's level?


### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

stanislavkozlovski committed:
Add BROKER_LOGGER config types

Adds the BROKER_LOGGER ConfigResource type
Adds the DYNAMIC_BROKER_LOGGER_CONFIG ConfigSource

stanislavkozlovski committed:
Add Log4jController#loggers

Also refactors the class to be public

stanislavkozlovski committed:
Update DescribeConfigs to work with broker logger configs

stanislavkozlovski committed:
Implement IncrementalAlterConfigs for broker loggers

Adds a LogLevelConfig public class, which specifies all the supported log levels
Adds a test which makes sure the new API uses the AlterConfigs Cluster resource
Adds exhaustive tests for the IncrementalAlterConfigs change

stanislavkozlovski committed:
ConfigCommand: Add broker-logger validations

Should only support --bootstrap-server and not support "entity-default"
Adds new EntityType object class

stanislavkozlovski committed:
ConfigCommand: Add broker logger functionality and tests

stanislavkozlovski commented on clients/src/main/java/org/apache/kafka/common/config/LogLevelConfig.java:
```diff
@@ -0,0 +1,60 @@
+    /**
+     * The <code>ALL</code> has the lowest possible rank and is intended to turn on all logging.
+     */
+    public static final String ALL_LOG_LEVEL = "ALL";
```
In reality, `TRACE` does the same job. I don't have a strong opinion whether we should support this or not.

gwenshap replied:
I'd prefer to leave this out. Basically because the rest are quite standard and this one is less expected.

stanislavkozlovski replied:
Agreed. I did not know about this log level until I dove into the log4j code

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -282,50 +282,69 @@ object ConfigCommand extends Config {
     val entityTypes = opts.options.valuesOf(opts.entityType).asScala
     if (entityTypes.size != 1)
       throw new IllegalArgumentException("Exactly one --entity-type must be specified with --bootstrap-server")
-    if (entityTypes.head != ConfigType.Broker)
```
this validation is now done in checkArgs()

stanislavkozlovski committed:
Minor polish up

stanislavkozlovski committed:
Fix header issue

stanislavkozlovski commented:
There is an issue with the failing integration tests. They modify the log levels and do not set them back. This results in the next log level test to get modified log levels it doesn't expect and thus fail.

I am not coming up with a clean way to solve this, as altering back to the ROOT logger would change the expected value. I think I might go with using different loggers for the tests so they don't conflict

stanislavkozlovski committed:
Use different loggers in integration tests to ensure test isolation

gwenshap commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeConfigsResponse.java:
```diff
@@ -179,7 +179,8 @@ public boolean isReadOnly() {
         STATIC_BROKER_CONFIG((byte) 4),
-        DEFAULT_CONFIG((byte) 5);
+        DEFAULT_CONFIG((byte) 5),
+        DYNAMIC_BROKER_LOGGER_CONFIG((byte) 6);
```
Nit: Default should be last (just like in switch statements and such, it is a useful convention)

stanislavkozlovski replied:
Agreed. I thought it was ordered by ID. We need to keep DEFAULT_CONFIG=5 for old clients, right?

stanislavkozlovski replied:
Ok I learned through test failures that that's not the ID but the ordinal. We should change them and code shouldn't depend on those anyway

gwenshap replied:
Wait, now I'm confused. I re-checked the code based on your comment and it does look like those numbers are the IDs that we are returning to clients, so changing the numbers break our protocol.  (looking at the `forId` method).
What did your tests reveal? Where does the actual ID come from?

Also maybe a thing to keep in mind (and document): It looks like older adminclient talking to a new broker may see the config source as "UNKNOWN CONFIG". I don't think it is a big deal, but just something to be aware of.

stanislavkozlovski replied:
You're 100% right. This ID is actually an index.

The reason my test failed was because I had the code like this:
```
public enum ConfigSource {
        UNKNOWN_CONFIG((byte) 0),
        TOPIC_CONFIG((byte) 1),
        DYNAMIC_BROKER_CONFIG((byte) 2),
        DYNAMIC_DEFAULT_BROKER_CONFIG((byte) 3),
        STATIC_BROKER_CONFIG((byte) 4),
        DYNAMIC_BROKER_LOGGER_CONFIG((byte) 6), -- NEW
        DEFAULT_CONFIG((byte) 5);
```
in order to keep the convention of defaults at the bottom. The test would fail because it would get a byte ID of 6 and get values[6], which is DEFAULT_CONFIG and not the expected dynamic broker logger.

So it turns out we can't keep the convention here :(

stanislavkozlovski replied:
An old AdminClient talking to a new broker should never see this config source at all because it would never request the broker loggers configs

gwenshap commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1888,6 +1889,9 @@ void handleFailure(Throwable throwable) {
             case DEFAULT_CONFIG:
                 configSource = ConfigEntry.ConfigSource.DEFAULT_CONFIG;
                 break;
+            case DYNAMIC_BROKER_LOGGER_CONFIG:
```
Nit: Should probably be places above DEFAULT_CONFIG, it is customary to place the default case last.

gwenshap commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -282,50 +282,69 @@ object ConfigCommand extends Config {
     try {
       if (opts.options.has(opts.alterOpt))
-        alterBrokerConfig(adminClient, opts, entityName)
+        alterBrokerConfig(adminClient, opts, entityTypes.head, entityName)
```
leaving it here because I can't find a more relevant location:
`opt.entityType` lists the types of entities, since you added a new entity, you should add it there (so it will show up when we show the CLI help). I think this is necessary in few other places in `ConfigCommandOptions` class, like in add-config.

stanislavkozlovski replied:
Agreed. I need to manually play around with the command

stanislavkozlovski replied:
> opt.entityType lists the types of entities, since you added a new entity, you should add it there (so it will show up when we show the CLI help). I think this is necessary in few other places in ConfigCommandOptions class, like in add-config.

I don't think I understand - the only place I found that I can add it such that it comes out in the CLI help is the description of the --entity-type option:
```
val entityType = parser.accepts("entity-type", "Type of entity (topics/clients/users/brokers/broker-loggers)")
```
✅ added it there

Additionally, I played around with the command, here is some output:
```
./kafka-configs.sh --bootstrap-server localhost:9092 --describe --entity-type broker-loggers
At least one of --entity-name or --entity-default must be specified with --bootstrap-server
```
```
./kafka-configs.sh --bootstrap-server localhost:9092 --describe --entity-type broker-loggers --entity-default
Error while executing config command with args '--bootstrap-server localhost:9092 --describe --entity-type broker-loggers --entity-default'
java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.InvalidRequestException: Broker id must not be empty
```
Hah - whoops. We don't have validation for describe with entity-default. Adding it ✅ 

--------------------------

Ugh, I get the following error locally on the broker (when the command passes validation):
```
[2019-06-24 10:50:23,401] ERROR [Admin Manager on Broker 0]: Error processing describe configs request for resource ConfigResource(type=BROKER_LOGGER, name='0') (kafka.server.AdminManager)
java.lang.NoClassDefFoundError: kafka/utils/Log4jController$
	at kafka.server.AdminManager.$anonfun$describeConfigs$1(AdminManager.scala:377)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:237)
	at scala.collection.immutable.Map$Map1.foreach(Map.scala:128)
	at scala.collection.TraversableLike.map(TraversableLike.scala:237)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:230)
	at scala.collection.AbstractTraversable.map(Traversable.scala:108)
	at kafka.server.AdminManager.describeConfigs(AdminManager.scala:322)
	at kafka.server.KafkaApis.handleDescribeConfigsRequest(KafkaApis.scala:2298)
	at kafka.server.KafkaApis.handle(KafkaApis.scala:161)
	at kafka.server.KafkaRequestHandler.run(KafkaRequestHandler.scala:69)
	at java.base/java.lang.Thread.run(Thread.java:844)
```
Any advice on how to build the project such that this works? I had done `./gradlew clean` prior

gwenshap replied:
I didn't have any issue with the classes... I just build and it seemed to work. Weird! 

I did have other odd issues. I have one broker and `broker.id=0`

When I run:
`bin/kafka-configs.sh --bootstrap-server localhost:9092 --describe --entity-type broker-loggers --entity-default --entity-name 1` 
the command just hangs. Not sure if this is expected behavior but seems unfriendly?

When I run:
`bin/kafka-configs.sh --bootstrap-server localhost:9092 --describe --entity-type broker-loggers --entity-default --entity-name 0` I get all the dynamic configs for the broker, but not the loggers:
```
Configs for broker 0 are:
  org.apache.zookeeper.CreateMode=null sensitive=false synonyms={}
  kafka.coordinator.transaction.ProducerIdManager=null sensitive=false synonyms={}
  org.apache.kafka.common.utils.AppInfoParser=null sensitive=false synonyms={}
  kafka.utils.KafkaScheduler=null sensitive=false synonyms={}
  kafka.log.ProducerStateManager=null sensitive=false synonyms={}
...
```

Any idea?

gwenshap replied:
Regarding the CLI help. I think this is indeed the only place. I thought we may need something under "entity-name" and "entity-default", but looks like this isn't necessary.

The one thing I'm less sure of - how will users know that the right entity name for broker-logger is the broker-id (and not the name of a logger for instance)?

stanislavkozlovski replied:
**re: right entity name**
I thought about that too but I couldn't figure out whether adding broker id to the entity-name again would be better.
a) keep it as it is
```
parser.accepts("entity-name", "Name of entity (topic name/client id/user principal name/broker id)")
```
b) add broker id
```
parser.accepts("entity-name", "Name of entity (topic name/client id/user principal name/broker id/broker id)")
```


Here's another thing we can do - I'll add validation that the entity-name is an integer when a broker is given. 

**re: command hanging**
I got this behavior too when testing out a separate config - this is because you are giving it an invalid broker ID (1) and the admin client starts issuing requests. It eventually times out.

```
Error while executing config command with args '--bootstrap-server localhost:9092 --describe --entity-type broker-loggers --entity-default --entity-name 1'
java.util.concurrent.TimeoutException
	at org.apache.kafka.common.internals.KafkaFutureImpl$SingleWaiter.await(KafkaFutureImpl.java:108)
	at org.apache.kafka.common.internals.KafkaFutureImpl.get(KafkaFutureImpl.java:272)
	at kafka.admin.ConfigCommand$.brokerLoggerConfigs(ConfigCommand.scala:374)
	at kafka.admin.ConfigCommand$.describeBrokerConfig(ConfigCommand.scala:344)
	at kafka.admin.ConfigCommand$.processBrokerConfig(ConfigCommand.scala:288)
	at kafka.admin.ConfigCommand$.main(ConfigCommand.scala:82)
	at kafka.admin.ConfigCommand.main(ConfigCommand.scala)
```
Diving into it more there doesn't seem to be an easy way to fix this - when we can't find a node by its ID in our current metadata, we [request an update](https://github.com/apache/kafka/blob/5216839d8bcda5c00825a0f4be866a525d948da8/clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java#L545) until the we either time out or the [provide method returns non-null](https://github.com/apache/kafka/blob/5216839d8bcda5c00825a0f4be866a525d948da8/clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java#L897).
Testing this locally, I see my AdminClient creates 7.8 requests per second. This is quite the needless load on the cluster - it respects the retry.backoff.ms but maybe it is better to have an exponential backoff of sorts (like this [conference talk suggests](https://www.youtube.com/watch?v=NHNZab_AzZo&t=18m20s), for instance).
I opened https://issues.apache.org/jira/browse/KAFKA-8607 about it

**re: getting dynamic broker**
The output you gave is the broker loggers. I am getting the same output too. These are not the dynamic configs


Thanks for the thorough testing - I appreciate it. It's funny to see us hitting the same things  and I see we both agree CLI quality is very important

gwenshap commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -520,30 +545,48 @@ object ConfigCommand extends Config {
     }
   }
+
+  object EntityType {
```
Isn't it redundant to have both ConfigType and EntityType? Any reason not to add the new type and methods to the original? In addition to reducing duplication (imagine the future intern asking which one to use when), it will also reduce the size of this PR.

stanislavkozlovski replied:
My thinking was the following:
`ConfigType` is defined in `kafka.server` and is related to the back-end functionality - it is the names of the zNodes. The javadoc says this as well:
```
/**
 * Represents all the entities that can be configured via ZK
 */
```
The new config type I'm adding called `"broker-loggers"` is only used in the front-end - that tool is the only class that uses it. I thought it would be more confusing if we had it in kafka.server rather than just define it in the command itself.
Now, when it comes to defining it in the command we could just have a separate variable called `brokerLoggerConfigType` but I also wanted to add two lists for validating zk-only allowed configs and broker-only allowed configs:
```scala
    val zkSupportedTypes = Seq(Topic, Client, User, Broker)
    val brokerSupportedTypes = Seq(Broker, BrokerLogger)
```
Those make sense to be in the back-end though.

I also thought that coupling the front-end flags with the back-end is more brittle, even if we are super unlikely to ever change a zNode path.
I don't think it'll be too confusing for new people as they can see that the one is defined and used entirely in the tools class whereas the other `ConfigType` is used throughout.

Regardless of the wall of text, I don't feel strongly about this :D. Let me know what you think after I explained my reasoning

gwenshap replied:
In terms of reasoning:
- Both entities are in the backend, so I don't think there is coupling of front end and back end. ConfigCommand and DynamicConfigManager are both inside core Kafka module. If ConfigCommand was in client module, I'd ask you to create ConfigType in `common` and make sure they both use it. It is literally shared information - and all types of dynamic configs will remain long after ZK is gone.
- From what I see, this isn't a public API, so the Javadoc can change :) There's no real contract around this as far as I can tell.
- I liked the two lists for ZK and not ZK. It cleaned the validation nicely. They can even be in ConfigCommand, they just don't require EntityType...
- I don't think it is confusing. I just think it is redundant. And you had to make larger changes to add the new type, which always add a bit of risk.

gwenshap commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -1819,6 +1819,123 @@ class AdminClientIntegrationTest extends IntegrationTestHarness with Logging {
+
+  @Test
+  def testDescribeConfigsForLog4jLogLevels(): Unit = {
+    createTopic(topic, numPartitions = 2, replicationFactor = brokerCount)
```
why are you creating a topic?

stanislavkozlovski replied:
https://en.wikipedia.org/wiki/Copy-and-paste_programming

gwenshap commented on core/src/test/scala/unit/kafka/admin/ConfigCommandTest.scala:
```diff
@@ -228,6 +239,26 @@ class ConfigCommandTest extends ZooKeeperTestHarness with Logging {
   }
 
+  @Test
+  def shouldAddBrokerLoggerConfig(): Unit = {
```
Did you consider also testing describe/alter on all brokers? (i.e. when entity name is empty)?
It is supported for broker configs, so either it works for logs too, or we need to document (and have an explicit error) that it doesn't.

stanislavkozlovski replied:
We do have an explicit error:
```scala
if (entityTypeVals.contains(EntityType.User) || entityTypeVals.contains(EntityType.Client) || entityTypeVals.contains(EntityType.Broker)) {
          if (!options.has(entityName) && !options.has(entityDefault))
            throw new IllegalArgumentException("--entity-name or --entity-default must be specified with --alter of users, clients or brokers")
        } else if (!options.has(entityName))
            throw new IllegalArgumentException(s"--entity-name must be specified with --alter of $entityTypeVals")
```
the `throw new IllegalArgumentException(s"--entity-name must be specified with --alter of $entityTypeVals")` should catch this.
great call on having a test case - adding one

gwenshap commented on core/src/test/scala/unit/kafka/admin/ConfigCommandTest.scala:
```diff
@@ -228,6 +239,26 @@ class ConfigCommandTest extends ZooKeeperTestHarness with Logging {
+  def shouldRaiseInvalidConfigurationExceptionWhenAddingInvalidBrokerLoggerConfig(): Unit = {
+    val node = new Node(1, "localhost", 9092)
+    // we want to alter kafka.log.LogCleaner, kafka.server.ReplicaManager and kafka.server.KafkaApi
+    // yet, we only get one logger returned
```
I don't think I understand this comment or this test... what is invalid in this config? If we expect an exception, why does the comment mention only one logger returned?

stanislavkozlovski replied:
I clarified the comment better:
```
// verifyAlterBrokerLoggerConfig tries to alter kafka.log.LogCleaner, kafka.server.ReplicaManager and kafka.server.KafkaApi
    // yet, we make it so DescribeConfigs returns only one logger, implying that kafka.server.ReplicaManager and kafka.log.LogCleaner are invalid
```

it's kind of tied to the implementation of `verifyAlterBrokerLoggerConfig`

gwenshap replied:
if it doesn't throw an exception, the name is pretty misleading. maybe "shouldNotReturnLoggerWhenAddingInvalidLogger"?

gwenshap commented on core/src/main/scala/kafka/utils/Log4jController.scala:
```diff
@@ -22,44 +22,78 @@ import java.util.Locale
-  def getLoggers = {
-    val lst = new util.ArrayList[String]()
-    lst.add("root=" + existingLogger("root").getLevel.toString)
+object Log4jController {
```
As far as I can see, this doesn't have to be synchronized because it is a wrapper on log4j which is thread-safe. https://logging.apache.org/log4j/1.2/faq.html#a1.7

gwenshap replied:
and static methods look fine to me for same reason.

gwenshap commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -356,8 +357,16 @@ class AdminManager(val config: KafkaConfig,
 
+          case ConfigResource.Type.BROKER_LOGGER =>
+            if (resource.name == null || resource.name.isEmpty)
+              throw new InvalidRequestException("Broker id must not be empty")
```
why not support this, like we do for broker configs?

stanislavkozlovski replied:
What does a lack of broker_id do for broker configs? It returns all the cluster-wide default configurations from ZK.
With broker loggers, we don't have a cluster-wide default configuration - only the locally-modified one made from the `log4j.properties`. Further, I don't think log4j gives us a way to get the default configs prior to any modification. I don't see what we can return here

gwenshap replied:
Sorry, it was my misunderstanding of what the empty broker_id returns.

gwenshap commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -504,6 +532,32 @@ class AdminManager(val config: KafkaConfig,
     }.toMap
   }
 
+  private def validateLogLevelConfigs(alterConfigOps: List[AlterConfigOp]): Unit = {
```
I like this validation. Comprehensive and still readable.

gwenshap commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -1819,6 +1819,123 @@ class AdminClientIntegrationTest extends IntegrationTestHarness with Logging {
   }
+
+  @Test
+  def testDescribeConfigsForLog4jLogLevels(): Unit = {
```
It isn't clear to me if:
1. We are testing returning a log level configuration set by log4j.properties file?
2. What would be the config source in that case

stanislavkozlovski replied:
That is what we are testing, yes. This is a bit of a sanity check to see it behaves as expected and also serves as a bit of test-documentation such that anybody reading the code can see a quick example of the API.

Regarding the config source, it should (and is, as tested) be the DYNAMIC_BROKER_LOGGER_CONFIG. I found this out during implementation and mentioned in the vote thread:
```
During implementation, I noticed a couple of holes in the KIP and did the following changes:
* add a new ConfigResource - DYNAMIC_BROKER_LOGGER_CONFIG
* DELETE operations in the IncrementalAlterConfigRequest will change the logger's level to the ROOT logger's currently-set level. (we previously thought we could return to the default - log4j does not seem to allow us to do that)
```

The key being I previously thought we had a way of knowing the default but we don't, therefore the server can't return a specific config source mentioning it's a default. Note that we can't infer the default from the root logger as that can be changed as well or the specific logger could be configured to use a non-default log level.

gwenshap replied:
So, I'm still not sure what will someone get if they try to describe the log level config for a broker that just started. Basically "null" or "unset"?

stanislavkozlovski replied:
`null`. They will always get `null` for a non-configured logger until they dynamically change it

stanislavkozlovski replied:
Changing it to return the root logger now

gwenshap commented on core/src/main/scala/kafka/utils/Log4jController.scala:
```diff
@@ -22,44 +22,78 @@ import java.util.Locale
       if (logger != null) {
-        val level =  if (logger != null) logger.getLevel else null
-        lst.add("%s=%s".format(logger.getName, if (level != null) level.toString else "null"))
+        val level = if (logger.getLevel != null) logger.getLevel.toString else "null"
```
I'm actually not sure if we should return root logger here... it could be confusing. On the other hand, null looks like an error. Maybe something like "not set"?

stanislavkozlovski replied:
I had left it as `null` since that is what our current Log4jController did and to keep the scope of interface changes low. It would be inconsistent if the MBean returned "null" and the API "not set"

Let's decide on this.
* "not set" - it's just a matter of naming and I agree this is more human readable. I agree we shouldn't leave it as "null" in any case
* root logger - I prefer this. I don't think this will be confusing. If anything, it will be more informative. In practice, if not set explicitly it should always reflect the root logger (unless some crazy hierarchy was configured in log4.properties I guess). As a user, if I see `null` or `not set` I would need to **know** and fetch the root logger.
I think that this is a better approach and I also think that we should apply this to the Log4j MBean as well.

I can update the KIP, mention in the vote thread and we can be good to go.
LMK what you think

gwenshap replied:
Yeah, you are right - root logger sounds good.

gwenshap reviewed (changes requested):
Lots of comments, but only a few are actually important.

I'm most curious about the decision not to support empty entity.

stanislavkozlovski committed:
Address PR comments:

- Move default values to the bottom for better convention
- Remove ALL log level
- Make some code comments clearer
- Add two tests for the new config command and entity-default behavior
- Check for entity-type/entity-default mandatory presence with --bootstrap-server in checkArgs()
- Further refactor validations in ConfigCommand (check supplied entity types first, validate --describe, broker-loggers and entity-name, validate --bootstrap-option and the need for --entity-type/--entity-default and --bootstrap-server)

stanislavkozlovski commented:
Scala 2.11 build failing with:
```
11:35:10 > Task :core:compileTestScala FAILED
11:35:10 Pruning sources from previous analysis, due to incompatible CompileSetup.
11:35:10 /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.11/core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:221: illegal start of simple expression
11:35:10   )
11:35:10   ^
11:35:10 /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.11/core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:1684: ')' expected but '}' found.
11:35:10 }
11:35:10 ^
11:35:10 two errors found
```

I think it's because of trailing commas which was added with 2.12.2 (https://docs.scala-lang.org/sips/trailing-commas.html and https://github.com/scala/scala/pull/5245)

stanislavkozlovski commented:
retest this please

stanislavkozlovski commented:
@mimaison @harshach do you have time to review this?

gwenshap commented:
Almost there! 3 concerns remain for me:
- Does re-ordering ConfigSource enum break the protocol? You said it did, and then said it doesn't, but it looks to me like it does?
- CLI behaved weirdly for both of us, but in different ways
- ConfigType vs EntityType. I still don't see why we need both. We have one "list of types of configs" that everyone can share. Whether or not some of them are in ZK doesn't require two lists IMO.

stanislavkozlovski committed:
Re-order Config Source enum

stanislavkozlovski committed:
Add brokerId validation to ConfigCommand

stanislavkozlovski committed:
Return the ROOT logger's level if an explicit log level is not set for a logger

We now return this from the Log4jController MBean as well

stanislavkozlovski committed:
Remove ConfigCommand#EntityType and continue using ConfigType

stanislavkozlovski committed:
Fix tests to obey broker entity name validation and add test for the validation

stanislavkozlovski commented:
I think I addressed everything outstanding

stanislavkozlovski commented:
The test failures are annoying to deal with. The crux of the issue is the following: because we're running multiple test classes using the AdminClientIntegration tests in the same JVM, changed log levels from one test leak onto the next, killing its expectations.

My naive approach is to set levels back to what they were at tearDown and I think that may be the way to go

stanislavkozlovski committed:
Clean up log levels in admin client integration test

stanislavkozlovski commented:
The latest commit has the solution to the test problems but it is not the cleanest I feel. Nevertheless, I found it hard to come up with anything better. Let's see if this works 🤞

stanislavkozlovski committed:
Correctly tear down loggers in integration test

stanislavkozlovski commented:
The test failures exposed an interesting gotcha. If we want to change the root logger's log level and delete the log levels of a logger in the same request (reset to root), we need to ensure the root logger's config entry is first such that it first gets changed and the subsequent deletes (reset to root) get set to the new root logger log level.

This is not the most pleasant-to-the-eye solution but at least it should be correct and not allow for any flakiness

stanislavkozlovski commented:
Strange, the test again failed with 
```
20:01:24 kafka.api.SaslSslAdminClientIntegrationTest > testIncrementalAlterConfigsForLog4jLogLevels FAILED
20:01:24     org.junit.ComparisonFailure: expected:<[FATAL]> but was:<[DEBUG]>
# at line 1850
```

I feel as if these tests are running concurrently and thus experiencing a race condition, I verified we are running with just one fork
```
17:59:14 + ./jenkins.sh -PmaxParallelForks=1 -PscalaVersion=2.12
```

stanislavkozlovski commented:
retest this please

stanislavkozlovski commented:
I'll run the tests locally as well with `./jenkins.sh -PmaxParallelForks=1 -PscalaVersion=2.12`

stanislavkozlovski commented:
I managed to reproduce this locally, will look further into it

stanislavkozlovski committed:
Reset logger level to null internally

This allows us to tell when a logger's level has been reset (should always return ROOT log level) or has been explicitly set to a level
Added a test to verify the behavior

stanislavkozlovski committed:
Ensure AdminClient requests go to the correct BROKER_LOGGER node

stanislavkozlovski committed:
Do not allow resetting the root logger's level

stanislavkozlovski commented:
Okay, this was tricky to reproduce. Let me illustrate the key issue via an example:
```
1. Assume ROOT logger == DEBUG
2. Change kafka.controller.KafkaController logger to INFO
3. Reset kafka.controller.KafkaController via AlterConfigOp.OpType.DELETE (resets it to the root logger - DEBUG)
4. Change ROOT logger to INFO
5. What is the kafka.controller.KafkaController level? - DEBUG, because we explicitly set it to debug in step 3
```
Good thing for the tests to make this bug clear. In step 5., we'd like the kafka.controller.KafkaController logger to have INFO level - the same as root.
Because we explicitly set it to the root logger's level at the time of resetting, upon a GET operation, we no way of knowing whether we want the current root logger's level or whatever we had set.

We basically need a placeholder value to tell us that this level is "reset" and should be the current ROOT logger's value. Let's make this be `null` - it is a permitted value when calling `setLevel()` (https://logging.apache.org/log4j/1.2/apidocs/)
I will change it to be null and add another test.

------------------------------

I found two other bugs that need fixing:
1. The KafkaAdminClient would send a BROKER_LOGGER ConfigResource to the least loaded node and not to the specific broker targeted. This was uncovered via the tests being flaky with the following error:
```
java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.InvalidRequestException: Unexpected broker id, expected 1, but received 0
```

This looks tricky to test explicitly in the KafkaAdminClientTest.java class without a major rewrite


2. We allowed users to reset the ROOT logger's level. We should just return an invalid request exception. Updated the KIP to mention it's not allowed.

---------------------------

I've added three new commits to address all of the issues

stanislavkozlovski committed:
Fix test flakiness - make log4j AuthorizerIntegrationTest reset a log level rather than set

stanislavkozlovski commented:
Test failures seem unrelated. Let's see if JDK 11/Scala 2.12 goes green 🙏

stanislavkozlovski commented:
retest this please

stanislavkozlovski commented:
@wicknicks suggested "unset" would be a better name for the reset operation, as it's more correct. 

We are "resetting" the logger to the root log level, but in reality we are unsetting it's log level so it naturally uses the loggers higher up in the hierarchy

@gwenshap I think I'll change the 10 or so occurrences in the code where we mention reset to unset. sound good?

gwenshap commented:
Sounds right to me!

stanislavkozlovski committed:
Rename some occurrences of reset to unset/remove in regards to log levels

stanislavkozlovski commented on core/src/main/scala/kafka/admin/ConfigCommand.scala:
```diff
@@ -349,6 +368,15 @@ object ConfigCommand extends Config {
+    */
+  private def brokerLoggerConfigs(adminClient: Admin, entityName: String): Seq[ConfigEntry] = {
+    val configResource = new ConfigResource(ConfigResource.Type.BROKER_LOGGER, entityName)
+    val configs = adminClient.describeConfigs(Collections.singleton(configResource)).all.get(30, TimeUnit.SECONDS)
```
The previous code had a 30 second timeout for describe and a 60 second for alter, so I kept with the convention

stanislavkozlovski committed:
Change "root" value to Log4jController.ROOT_LOGGER in tests

Also add a helpful space in an exception message :)

stanislavkozlovski commented:
Had to rebase with master due to a small conflict with https://github.com/apache/kafka/commit/2a133ba656ae550cc9ad4bcda0a9afb27b14f4e0 -- it was just some conflicts in altered method signatures due to my changes using the old JAdminClient interface. Now it uses the Admin interface

gwenshap commented:
@stanislavkozlovski This failure on Scala 2.13 looks a bit suspicious, no? 
`kafka.admin.ConfigCommandTest.shouldAddBrokerLoggerConfig`

stanislavkozlovski commented:
True. Unfortunately, I missed the build and it's deleted from Jenkins now, so I can't tell what the failure is.
I couldn't get a build running locally with Scala 2.13 (might it be because I'm using JDK 10?). I ran the test 50 times with Scala 2.12 and then ran ConfigCommandTest 50 times. No flakes whatsoever. Let's see if it will fail again

retest this please

wicknicks commented:
@stanislavkozlovski, regarding some of the open questions, here are some thoughts:

> Should we add synchronization to the Log4jController methods? - Seems like we don't get much value from it

Agree. Might not bring a lot of value here. 

> A logger which does not have a set value returns "null" (as seen in the unit tests). Should we just return the Root logger's level?

I have a PR here https://github.com/apache/kafka/pull/6069 and I chose to return effective log level [1], because a null is technically correct but not informative. If a user is trying to check what the log level for a class is because they can't find any logs from the class in the output, `null` will not be very helpful (which IMHO is the primary use case for get log level feature). Instead, effectiveLogLevel will give you the log level of some parent (could be the root logger) who level is being used. 

[1] https://github.com/apache/kafka/pull/6069/files#diff-c1e7b589e0448d4df0b329da9d8b1ff4R96

stanislavkozlovski commented:
@wicknicks thanks for the input.

With regards to returning the effective log level, it should always return the ROOT logger's level with our default log4j settings, correct?
I think it makes total sense to use that rather than return the root logger's level

stanislavkozlovski commented:
`kafka.admin.ConfigCommandTest.shouldAddBrokerLoggerConfig` failed again with Scala 2.13
Full error is:
```
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:87)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.junit.Assert.assertTrue(Assert.java:53)
	at kafka.admin.ConfigCommandTest$$anon$2.incrementalAlterConfigs(ConfigCommandTest.scala:409)
	at kafka.admin.ConfigCommand$.alterBrokerConfig(ConfigCommand.scala:332)
	at kafka.admin.ConfigCommandTest.verifyAlterBrokerLoggerConfig(ConfigCommandTest.scala:415)
	at kafka.admin.ConfigCommandTest.shouldAddBrokerLoggerConfig(ConfigCommandTest.scala:252)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:305)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:365)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:330)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:78)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:328)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:65)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:292)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:305)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:412)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at jdk.internal.reflect.GeneratedMethodAccessor110.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:118)
	at jdk.internal.reflect.GeneratedMethodAccessor109.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:175)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:157)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:404)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

Failing at this line - https://github.com/apache/kafka/pull/6903/files#diff-bd940023d2b011cfc185db98bd66d5e9R409


I managed to run locally with JDK 11 and Scala 2.13, yet I couldn't reproduce. I'll push a commit here to better understand the errr

stanislavkozlovski committed:
TEMP - Descriptive error message

stanislavkozlovski commented:
The error message was
```
java.lang.AssertionError: Expected expectedConfigOps=List(AlterConfigOp{opType=SET, configEntry=ConfigEntry(name=kafka.log.LogCleaner, value=DEBUG, source=UNKNOWN, isSensitive=false, isReadOnly=false, synonyms=[])}, AlterConfigOp{opType=DELETE, configEntry=ConfigEntry(name=kafka.server.ReplicaManager, value=, source=UNKNOWN, isSensitive=false, isReadOnly=false, synonyms=[])}, AlterConfigOp{opType=DELETE, configEntry=ConfigEntry(name=kafka.server.KafkaApi, value=, source=UNKNOWN, isSensitive=false, isReadOnly=false, synonyms=[])}) 
 got View(<not computed>)
```

stanislavkozlovski committed:
Try fixing the test through toList

stanislavkozlovski commented:
Awesome, this one passed Scala 2.13.

Retest this please

stanislavkozlovski committed:
Polish up test assertion

stanislavkozlovski commented:
Had to rebase with trunk

gwenshap commented:
LGTM. Merging.
